src/main/java/com/hivemq/client/internal/mqtt/handler/subscribe/MqttSubscriptionHandler.java

```diff
@@ -244,79 +227,75 @@ public void channelRead(final @NotNull ChannelHandlerContext ctx, final @NotNull
     }
 
     private void readSubAck(final @NotNull ChannelHandlerContext ctx, final @NotNull MqttSubAck subAck) {
-        final int packetIdentifier = subAck.getPacketIdentifier();
-        final MqttSubOrUnsubWithFlow.Stateful statefulSubOrUnsubWithFlow = pendingIndex.remove(packetIdentifier);
+        final MqttSubOrUnsubWithFlow subOrUnsubWithFlow = pendingIndex.remove(subAck.getPacketIdentifier());
 
-        if (statefulSubOrUnsubWithFlow == null) {
+        if (subOrUnsubWithFlow == null) {
             MqttDisconnectUtil.disconnect(ctx.channel(), Mqtt5DisconnectReasonCode.PROTOCOL_ERROR,
                     "Unknown packet identifier for SUBACK");
             return;
         }
-        if (!(statefulSubOrUnsubWithFlow instanceof MqttSubscribeWithFlow.Stateful)) {
+        if (!(subOrUnsubWithFlow instanceof MqttSubscribeWithFlow)) {
             MqttDisconnectUtil.disconnect(ctx.channel(), Mqtt5DisconnectReasonCode.PROTOCOL_ERROR,
                     "SUBACK received for an UNSUBSCRIBE");
             return;
         }
-        final MqttSubscribeWithFlow.Stateful statefulSubscribeWithFlow =
-                (MqttSubscribeWithFlow.Stateful) statefulSubOrUnsubWithFlow;
-        final MqttStatefulSubscribe subscribe = statefulSubscribeWithFlow.getMessage();
-        final MqttSubscriptionFlow<MqttSubAck> flow = statefulSubscribeWithFlow.getFlow();
+        final MqttSubscribeWithFlow subscribeWithFlow = (MqttSubscribeWithFlow) subOrUnsubWithFlow;
+        final MqttSubscriptionFlow<MqttSubAck> flow = subscribeWithFlow.getFlow();
 
         final ImmutableList<Mqtt5SubAckReasonCode> reasonCodes = subAck.getReasonCodes();
-        final boolean countNotMatching = subscribe.stateless().getSubscriptions().size() != reasonCodes.size();
+        final boolean countNotMatching = subscribeWithFlow.subscribe.getSubscriptions().size() != reasonCodes.size();
         final boolean allErrors = MqttCommonReasonCode.allErrors(subAck.getReasonCodes());
 
-        incomingPublishFlows.subAck(subscribe, subAck, statefulSubscribeWithFlow.getPublishFlow());
+        incomingPublishFlows.subAck(subscribeWithFlow.subscribe, subscribeWithFlow.subscriptionIdentifier, reasonCodes);
```

Comment:
Shouldnt there be some circuit breaking around the whole suback handling? This leads to a lot of recursion further down the callstack and the input for this comes from outisde the system instead of from the inside.

Reply:
This does not lead to recursion as the tree inside does not use recursion.

Reply:
Thats the reason why the code of the SubscribedPublishFlowTree is pretty complicated

Reply:
Ok let me check again.

Reply:
Yes I was mistaken. (As expected I might add ;)